File: Code Base Backend/app/services/jobs.py
"""
Background Jobs Service using Trigger.dev

Handles asynchronous tasks:
- Sending notifications (email, push)
- Processing alerts
- Data synchronization
"""

from typing import Dict, Any, Optional, List
import httpx
from datetime import datetime

from app.core.config import settings


class TriggerDevService:
    """Trigger.dev background jobs service."""

    def __init__(self):
        self.api_key = getattr(settings, 'TRIGGER_API_KEY', None)
        self.api_url = getattr(settings, 'TRIGGER_API_URL', 'https://api.trigger.dev')
        self.client = httpx.AsyncClient(
            base_url=self.api_url,
            headers={
                'Authorization': f'Bearer {self.api_key}',
                'Content-Type': 'application/json'
            } if self.api_key else {}
        )

    async def trigger_job(
        self,
        job_id: str,
        payload: Dict[str, Any],
        options: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Trigger a background job.

        Args:
            job_id: The job identifier (e.g., 'send-notification', 'process-alert')
            payload: Job data
            options: Job options (delay, timeout, etc.)
        """
        if not self.api_key:
            # Development mode - log and return mock response
            print(f"[Trigger.dev Mock] Job: {job_id}")
            print(f"[Trigger.dev Mock] Payload: {payload}")
            return {
                "id": f"mock-{job_id}-{datetime.utcnow().timestamp()}",
                "status": "queued",
                "mock": True
            }

        try:
            response = await self.client.post(
                f'/api/v1/jobs/{job_id}/trigger',
                json={
                    "payload": payload,
                    "options": options or {}
                }
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"Trigger.dev error: {e}")
            return {"error": str(e), "status": "failed"}

    # ==================== Notification Jobs ====================

    async def send_email(
        self,
        to: str,
        subject: str,
        body: str,
        html_body: Optional[str] = None,
        priority: str = "normal"
    ) -> Dict[str, Any]:
        """Send email notification."""
        return await self.trigger_job(
            "send-email",
            {
                "type": "email",
                "to": to,
                "subject": subject,
                "body": body,
                "html_body": html_body,
                "priority": priority,
                "timestamp": datetime.utcnow().isoformat()
            }
        )

    async def send_push_notification(
        self,
        user_id: str,
        title: str,
        body: str,
        data: Optional[Dict[str, Any]] = None,
        priority: str = "normal"
    ) -> Dict[str, Any]:
        """Send push notification."""
        return await self.trigger_job(
            "send-push",
            {
                "type": "push",
                "user_id": user_id,
                "title": title,
                "body": body,
                "data": data or {},
                "priority": priority,
                "timestamp": datetime.utcnow().isoformat()
            }
        )

    # ==================== Alert Jobs ====================

    async def process_alert(
        self,
        alert_id: str,
        title: str,
        message: str,
        priority: str,
        target_users: List[str],
        target_roles: List[str]
    ) -> Dict[str, Any]:
        """Process and distribute alert to target users."""
        return await self.trigger_job(
            "process-alert",
            {
                "alert_id": alert_id,
                "title": title,
                "message": message,
                "priority": priority,
                "target_users": target_users,
                "target_roles": target_roles,
                "timestamp": datetime.utcnow().isoformat()
            },
            options={"priority": "high" if priority == "critical" else "normal"}
        )

    async def send_critical_vitals_alert(
        self,
        patient_id: str,
        patient_name: str,
        bed_number: str,
        vitals_message: str,
        target_roles: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """Send critical vitals alert - high priority."""
        if target_roles is None:
            target_roles = ["doctor", "nurse", "admin"]

        return await self.trigger_job(
            "critical-vitals-alert",
            {
                "patient_id": patient_id,
                "patient_name": patient_name,
                "bed_number": bed_number,
                "message": vitals_message,
                "target_roles": target_roles,
                "timestamp": datetime.utcnow().isoformat()
            },
            options={"priority": "critical", "timeout": 300}
        )

    async def send_police_case_alert(
        self,
        case_id: str,
        patient_name: str,
        case_type: str,
        reported_by: str
    ) -> Dict[str, Any]:
        """Send police case alert to admins."""
        return await self.trigger_job(
            "police-case-alert",
            {
                "case_id": case_id,
                "patient_name": patient_name,
                "case_type": case_type,
                "reported_by": reported_by,
                "target_roles": ["admin"],
                "timestamp": datetime.utcnow().isoformat()
            },
            options={"priority": "high"}
        )

    # ==================== Scheduled Jobs ====================

    async def schedule_reminder(
        self,
        reminder_id: str,
        user_id: str,
        message: str,
        schedule_at: str
    ) -> Dict[str, Any]:
        """Schedule a reminder."""
        return await self.trigger_job(
            "schedule-reminder",
            {
                "reminder_id": reminder_id,
                "user_id": user_id,
                "message": message
            },
            options={"delay": schedule_at}
        )

    async def schedule_patient_followup(
        self,
        patient_id: str,
        doctor_id: str,
        followup_date: str,
        notes: str
    ) -> Dict[str, Any]:
        """Schedule patient follow-up reminder."""
        return await self.trigger_job(
            "patient-followup",
            {
                "patient_id": patient_id,
                "doctor_id": doctor_id,
                "followup_date": followup_date,
                "notes": notes
            },
            options={"delay": followup_date}
        )

    # ==================== Batch Operations ====================

    async def send_batch_notifications(
        self,
        notifications: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Send batch notifications."""
        return await self.trigger_job(
            "batch-notifications",
            {
                "notifications": notifications,
                "count": len(notifications),
                "timestamp": datetime.utcnow().isoformat()
            }
        )

    async def cleanup_old_data(
        self,
        days_old: int = 90
    ) -> Dict[str, Any]:
        """Cleanup old data (scheduled job)."""
        return await self.trigger_job(
            "cleanup-old-data",
            {
                "days_old": days_old,
                "timestamp": datetime.utcnow().isoformat()
            }
        )

    async def close(self):
        """Close HTTP client."""
        await self.client.aclose()


# Global instance
trigger_service = TriggerDevService()

File: Code Base Backend/trigger/jobs.ts
/**
 * Trigger.dev Job Definitions
 *
 * These jobs run in the background and handle:
 * - Sending notifications
 * - Processing alerts
 * - Scheduled tasks
 */

import { Trigger } from "@trigger.dev/sdk";
import { Resend } from "@trigger.dev/resend";

// Initialize Trigger client
const client = new Trigger({
  id: "er-command-center",
  apiKey: process.env.TRIGGER_API_KEY!,
  apiUrl: process.env.TRIGGER_API_URL,
});

// Initialize integrations
const resend = new Resend({
  id: "resend",
  apiKey: process.env.RESEND_API_KEY!,
});

// ==================== Notification Jobs ====================

/**
 * Send Email Notification
 */
client.defineJob({
  id: "send-email",
  name: "Send Email Notification",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "email.send",
  }),
  integrations: {
    resend,
  },
  run: async (payload, io, ctx) => {
    const { to, subject, body, html_body, priority } = payload;

    await io.logger.info(`Sending email to ${to}`);

    // Send email via Resend
    const result = await io.resend.emails.send("send-email", {
      from: "ER Command Center <notifications@ercommandcenter.com>",
      to: [to],
      subject: subject,
      text: body,
      html: html_body || body,
    });

    await io.logger.info(`Email sent: ${result.id}`);

    return {
      success: true,
      messageId: result.id,
    };
  },
});

/**
 * Send Push Notification
 */
client.defineJob({
  id: "send-push",
  name: "Send Push Notification",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "push.send",
  }),
  run: async (payload, io, ctx) => {
    const { user_id, title, body, data, priority } = payload;

    await io.logger.info(`Sending push notification to user ${user_id}`);

    // TODO: Integrate with Firebase Cloud Messaging or OneSignal
    // For now, log the notification
    await io.logger.info(`Push: ${title} - ${body}`);

    return {
      success: true,
      messageId: `push-${Date.now()}`,
    };
  },
});

// ==================== Alert Processing ====================

/**
 * Process Alert and Notify Users
 */
client.defineJob({
  id: "process-alert",
  name: "Process Alert",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "alert.process",
  }),
  run: async (payload, io, ctx) => {
    const { alert_id, title, message, priority, target_users, target_roles } = payload;

    await io.logger.info(`Processing alert ${alert_id} for roles: ${target_roles.join(", ")}`);

    // TODO: Query database for users with target_roles
    // TODO: Send push notifications to all target users
    // TODO: For critical alerts, also send SMS

    if (priority === "critical") {
      await io.logger.warn(`CRITICAL ALERT: ${title}`);
      // Send SMS to on-duty staff
    }

    return {
      success: true,
      alert_id,
      notified_users: target_users.length,
    };
  },
});

/**
 * Critical Vitals Alert
 */
client.defineJob({
  id: "critical-vitals-alert",
  name: "Critical Vitals Alert",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "vitals.critical",
  }),
  run: async (payload, io, ctx) => {
    const { patient_id, patient_name, bed_number, message, target_roles } = payload;

    await io.logger.warn(`CRITICAL VITALS: ${patient_name} (${bed_number})`);

    // Send immediate notifications to all nurses and doctors on duty
    // This is high priority - use SMS + Push

    return {
      success: true,
      patient_id,
      alert_sent: true,
    };
  },
});

/**
 * Police Case Alert
 */
client.defineJob({
  id: "police-case-alert",
  name: "Police Case Alert",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "police-case.created",
  }),
  run: async (payload, io, ctx) => {
    const { case_id, patient_name, case_type, reported_by } = payload;

    await io.logger.info(`Police case ${case_id} created for ${patient_name}`);

    // Notify all admin users
    // Send email and push notification

    return {
      success: true,
      case_id,
    };
  },
});

// ==================== Scheduled Jobs ====================

/**
 * Patient Follow-up Reminder
 */
client.defineJob({
  id: "patient-followup",
  name: "Patient Follow-up Reminder",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "followup.reminder",
  }),
  run: async (payload, io, ctx) => {
    const { patient_id, doctor_id, followup_date, notes } = payload;

    await io.logger.info(`Follow-up reminder for patient ${patient_id}`);

    // Send reminder to doctor
    // Send reminder to patient (if contact info available)

    return {
      success: true,
      patient_id,
      doctor_id,
    };
  },
});

/**
 * Batch Notifications
 */
client.defineJob({
  id: "batch-notifications",
  name: "Batch Notifications",
  version: "1.0.0",
  trigger: client.trigger.event({
    name: "notifications.batch",
  }),
  run: async (payload, io, ctx) => {
    const { notifications, count } = payload;

    await io.logger.info(`Processing ${count} notifications`);

    // Process notifications in batches
    for (const notification of notifications) {
      if (notification.type === "sms") {
        await io.sendEvent("send-sms", notification);
      } else if (notification.type === "email") {
        await io.sendEvent("send-email", notification);
      } else if (notification.type === "push") {
        await io.sendEvent("send-push", notification);
      }
    }

    return {
      success: true,
      processed: count,
    };
  },
});

/**
 * Cleanup Old Data (runs daily)
 */
client.defineJob({
  id: "cleanup-old-data",
  name: "Cleanup Old Data",
  version: "1.0.0",
  trigger: client.trigger.scheduled({
    cron: "0 2 * * *", // Run at 2 AM daily
  }),
  run: async (payload, io, ctx) => {
    await io.logger.info("Running data cleanup job");

    // TODO: Delete old audit logs
    // TODO: Archive old patient records
    // TODO: Clean up expired sessions

    return {
      success: true,
      cleaned_at: new Date().toISOString(),
    };
  },
});

export { client };
